# 6. Dane wejściowe, rekursja, obsługa plików

Szymon Dziubak

### Zadania

__Zadanie 1.__ <br>
Napisz funckję z użyciem danych wejściowych do liczenia BMI. <br>
Wowłaj funkcję dla wzrostu = 180cm i wagi = 80kg.

In [39]:
waga = input("Podaj wage ")
wzrost = input("Podaj wzrost w cm ")

def BMI (waga, wzrost):
    return (waga)/(wzrost/100)**2

BMI(int(waga), int(wzrost))

Podaj wage 80
Podaj wzrost w cm 180


24.691358024691358

__Zadanie 2.__ <br>
Napisz funkcję zwracającą konkretny element ciągu fibonacciego za pomocą pętli, rekursji i rekursji ogonowej

In [23]:
#przypomnienie - funkcja z pętlą wyglądała tak:
def ciag_fibonacciego(ilosc_elementow):
    x_0 = ''
    x_1 = 0
    x_2 = 1
    print(f"{x_1}\n{x_2}")
    for i in range(3, ilosc_elementow+1):
        x_0 = x_1
        x_1 = x_2
        x_2 = x_0 + x_1
        print(x_2)
#trzeba teraz przerobić tą funkcję aby zwracała konkretny element
def ciag_fibonacciego_wyraz(n):
    x_0 = ''
    x_1 = 0
    x_2 = 1
    if n == 1:
        return x_1
    elif n==2:
        return x_2
    else:    
        for i in range(2, n+1):
            x_0 = x_1
            x_1 = x_2
            x_2 = x_0 + x_1
    return x_2

In [26]:
ciag_fibonacciego_wyraz(10)

55

In [49]:
def fibonacci_rekurencyjnie(n):
    if n <= 1:
        return n 
    return fibonacci_rekurencyjnie(n-1)+fibonacci_rekurencyjnie(n-2)

In [50]:
fibonacci_rekurencyjnie(10)

55

In [48]:
def fibonacci_rekurencyjnie_ogon(n, a=0, b=1):
    if n == 0:
        return a
    if n == 1:
        return b
    return  fibonacci_rekurencyjnie_ogon(n-1, b, a+b) #nowe n to n-1, nowe a to b a nowe b to a+b

In [51]:
fibonacci_rekurencyjnie_ogon(10)

55

__Zadanie 3.__ <br>
Sprawdź, która funkcja najszybciej zwróciła wynik (dla dużego wyrazu np. 40)

In [52]:
# porównajmy czas wykonania pętli, rekursji zwykłej i ogonowej
import time
n=40

start_petla = time.time()
ciag_fibonacciego_wyraz(n)
print(f"Zwykła petla trwała {time.time() - start_petla}")

start_zwykla = time.time()
fibonacci_rekurencyjnie(n)
print(f"Zwykła rekursja trwała {time.time() - start_zwykla}")

start_ogonowa = time.time()
fibonacci_rekurencyjnie_ogon(n)
print(f"Rekursja ogonowa trwała {time.time() - start_ogonowa}")

Zwykła petla trwała 0.0
Zwykła rekursja trwała 35.53845453262329
Rekursja ogonowa trwała 0.0


__Zadanie 4.__ <br>
W folderze pliki znajduje się plik atp6-sequences.fasta. Jest to plik z sekwencją genu atp6.<br> Spróbuj otworzyć plik i sprawdzić jego format.<br>
Źródło: https://ggoralski.gitlab.io/python-wprowadzenie/czesc_i/16-pliki/

__Zadanie 5.__ <br>
Napisz kod który odczyta plik linia po linii.

In [27]:
with open(sciezka+r"/6. Pliki/atp6-sequences.fasta", 'r') as plik:
    for l in plik:
        print(l, end = "")
    #linie = plik.readlines()
    #for linia in linie:
    #    print(linia)

>KU180473.1 Orobanche coerulescens clone G18 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial
CTGCTAACTCTCAGTTTGGTCCTACTTCTGATTCATTTTGTTACTAAAAAGGGAGGAGGAAACTCAGTAC
CAAATGCTTGGCAATCCGTGGTAGAGTTTATTTATGATTTCGTGCTGAACCTGGTAAACGAACAAATAGG
GGGTCTTTCCGGAAATGTTAAACAAAAGTTTTTCCCTTGCATCTTGGTCACTTTTACTTTTTTGTTATTT
TGTAATCTTCAGGGTATGATACCTTATAGCTTCACAGTTACAAGTCATTTTCTCATTACTTTAGGTCTCT
CATTTTCTCTTTTTATTGGCATTACTATAGTGGGATTTCAAAGAAATGGGCTTCATTTTTTAAGCTTCTT
ATTACCCGCAGGAGTCCCACTGCCATTAGCACCTTTTTTAGTACTCCTTGAGCTAATTTCTTATTGTTTT
CGCGCATTAAGCTTAGGAATACGTTTATTTGCTAATATGATGGCCGGTCATAGTTTAGTAAAGATTTTAA
GTGGGTTCGCTTGGACTATGCTATGTATGAATGATCTTTTGTATTTTATAGGGGATCTTGGTCCTTTATT
TATAGTTCTTGCATTAACCGGTCTTGAATTAGGTGTAGCTATATCACAAGCTCATGTTTCTACGATCTCA
ATCTGTATTTAC

>KU180472.1 Orobanche cernua clone G15 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial
CTACTCACTCTCAGTTTGGTCCTACTTTTTGTTCATTTTGTTACTAAAAAGGGAGGAGGAAACTCAGTAC
CAAATGCTTTTCAATCCGTGTTAGAGCTTATTTATGATTTTGTGCCGAACCTGGTAAACGAACAAATAGG
TGGTCTT

__Zadanie 6.__ <br>
Opis sekwencji znajduje się w linii zaczynającej się od ">".<br>
Pozostałe linie to sekwencja.<br>
Puste linie mają długość 0 lub 1.<br>
Otwórz plik z dodaniem informacji, czy jest to opis, sekwencja czy pusta linia.<br>

In [30]:
with open(sciezka+r"/6. Pliki/atp6-sequences.fasta", 'rt') as plik:
    for l in plik:
        # Opis
        if l.startswith('>'):
            print("opis: ", end='')
        # Sekwencja
        elif len(l) > 1:
            print("sekwencja: ", end='')
        # Pusta
        else:
            print("pusta: ", end='')
        print(f"{l}", end='')

opis: >KU180473.1 Orobanche coerulescens clone G18 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial
sekwencja: CTGCTAACTCTCAGTTTGGTCCTACTTCTGATTCATTTTGTTACTAAAAAGGGAGGAGGAAACTCAGTAC
sekwencja: CAAATGCTTGGCAATCCGTGGTAGAGTTTATTTATGATTTCGTGCTGAACCTGGTAAACGAACAAATAGG
sekwencja: GGGTCTTTCCGGAAATGTTAAACAAAAGTTTTTCCCTTGCATCTTGGTCACTTTTACTTTTTTGTTATTT
sekwencja: TGTAATCTTCAGGGTATGATACCTTATAGCTTCACAGTTACAAGTCATTTTCTCATTACTTTAGGTCTCT
sekwencja: CATTTTCTCTTTTTATTGGCATTACTATAGTGGGATTTCAAAGAAATGGGCTTCATTTTTTAAGCTTCTT
sekwencja: ATTACCCGCAGGAGTCCCACTGCCATTAGCACCTTTTTTAGTACTCCTTGAGCTAATTTCTTATTGTTTT
sekwencja: CGCGCATTAAGCTTAGGAATACGTTTATTTGCTAATATGATGGCCGGTCATAGTTTAGTAAAGATTTTAA
sekwencja: GTGGGTTCGCTTGGACTATGCTATGTATGAATGATCTTTTGTATTTTATAGGGGATCTTGGTCCTTTATT
sekwencja: TATAGTTCTTGCATTAACCGGTCTTGAATTAGGTGTAGCTATATCACAAGCTCATGTTTCTACGATCTCA
sekwencja: ATCTGTATTTAC
pusta: 
opis: >KU180472.1 Orobanche cernua clone G15 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial
sekwencja: CTACTCACT

__Zadanie 7.__ <br>
Zauważ, że dla jednego opisu istnieje wiele linii sekwencji.
Teraz usuńmy ">" z opisu i połączmy linie sekwencje z jednego opisu w jedną.<br>
Podpowiedź: replace('>', '')

In [38]:
sekwencja = ''
with open(sciezka+r"/6. Pliki/atp6-sequences.fasta", 'r') as plik:
    for l in plik:
        if l.startswith('>'):
            opis = l.replace('>','').rstrip()
        elif len(l) > 1:
            sekwencja += l.rstrip()
        else:
            # Jeżeli mamy białą linię czyli koniec sekwencji to wyświetlamy opis i sekwencje
            print(f"opis: {opis} ")
            print(f"sekwencja: {sekwencja}")
            sekwencja = ''

opis: KU180473.1 Orobanche coerulescens clone G18 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial 
sekwencja: CTGCTAACTCTCAGTTTGGTCCTACTTCTGATTCATTTTGTTACTAAAAAGGGAGGAGGAAACTCAGTACCAAATGCTTGGCAATCCGTGGTAGAGTTTATTTATGATTTCGTGCTGAACCTGGTAAACGAACAAATAGGGGGTCTTTCCGGAAATGTTAAACAAAAGTTTTTCCCTTGCATCTTGGTCACTTTTACTTTTTTGTTATTTTGTAATCTTCAGGGTATGATACCTTATAGCTTCACAGTTACAAGTCATTTTCTCATTACTTTAGGTCTCTCATTTTCTCTTTTTATTGGCATTACTATAGTGGGATTTCAAAGAAATGGGCTTCATTTTTTAAGCTTCTTATTACCCGCAGGAGTCCCACTGCCATTAGCACCTTTTTTAGTACTCCTTGAGCTAATTTCTTATTGTTTTCGCGCATTAAGCTTAGGAATACGTTTATTTGCTAATATGATGGCCGGTCATAGTTTAGTAAAGATTTTAAGTGGGTTCGCTTGGACTATGCTATGTATGAATGATCTTTTGTATTTTATAGGGGATCTTGGTCCTTTATTTATAGTTCTTGCATTAACCGGTCTTGAATTAGGTGTAGCTATATCACAAGCTCATGTTTCTACGATCTCAATCTGTATTTAC
opis: KU180472.1 Orobanche cernua clone G15 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial 
sekwencja: CTACTCACTCTCAGTTTGGTCCTACTTTTTGTTCATTTTGTTACTAAAAAGGGAGGAGGAAACTCAGTACCAAATGCTTTTCAATCCGTGTTAGAGCTTATTTATGATTTTGTGCCGAACCTGGT

__Zadanie 8.__ <br>
Stwórzmy słownik, gdzie kluczem będzie opis a wartością sekwencja

In [41]:
sekwencje = {}
sekwencja = ''
with open(sciezka+r"/6. Pliki/atp6-sequences.fasta", 'r') as plik:
    for linia in plik:
        if linia.startswith('>'):
            opis = linia.replace('>','').rstrip()
        elif len(linia) > 1:
            sekwencja += linia.rstrip()
        else:
            sekwencje[opis] = sekwencja # Dodajemy do słownika
            sekwencja = '' # To będzie nowa sekwencja
print(sekwencje)

{'KU180473.1 Orobanche coerulescens clone G18 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial': 'CTGCTAACTCTCAGTTTGGTCCTACTTCTGATTCATTTTGTTACTAAAAAGGGAGGAGGAAACTCAGTACCAAATGCTTGGCAATCCGTGGTAGAGTTTATTTATGATTTCGTGCTGAACCTGGTAAACGAACAAATAGGGGGTCTTTCCGGAAATGTTAAACAAAAGTTTTTCCCTTGCATCTTGGTCACTTTTACTTTTTTGTTATTTTGTAATCTTCAGGGTATGATACCTTATAGCTTCACAGTTACAAGTCATTTTCTCATTACTTTAGGTCTCTCATTTTCTCTTTTTATTGGCATTACTATAGTGGGATTTCAAAGAAATGGGCTTCATTTTTTAAGCTTCTTATTACCCGCAGGAGTCCCACTGCCATTAGCACCTTTTTTAGTACTCCTTGAGCTAATTTCTTATTGTTTTCGCGCATTAAGCTTAGGAATACGTTTATTTGCTAATATGATGGCCGGTCATAGTTTAGTAAAGATTTTAAGTGGGTTCGCTTGGACTATGCTATGTATGAATGATCTTTTGTATTTTATAGGGGATCTTGGTCCTTTATTTATAGTTCTTGCATTAACCGGTCTTGAATTAGGTGTAGCTATATCACAAGCTCATGTTTCTACGATCTCAATCTGTATTTAC', 'KU180472.1 Orobanche cernua clone G15 ATPase subunit 6 (atp6) gene, partial cds; mitochondrial': 'CTACTCACTCTCAGTTTGGTCCTACTTTTTGTTCATTTTGTTACTAAAAAGGGAGGAGGAAACTCAGTACCAAATGCTTTTCAATCCGTGTTAGAGCTTATTTATGATTTTGTGCCGAACCTGGTAAACGAACAAATAGGTGGTCTTTCC

__Zadanie 9.__ <br>
Odczytaj plik MARKI_2019_05.csv <br>
Źródło: https://dane.gov.pl/pl/dataset/1643/resource/18173,marki-maj-2019-csv/table?page=1&per_page=20&q=&sort=

In [46]:
with open(sciezka+r"/6. Pliki/MARKI_2019_05.csv", 'r') as plik:
    for l in plik:
        print(l, end = "")
        

Marka|Liczba|
VOLKSWAGEN|15471|
OPEL|12410|
FORD|10590|
RENAULT|9683|
TOYOTA|9153|
AUDI|8248|
SKODA|7889|
BMW|7615|
MERCEDES-BENZ|6611|
PEUGEOT|6568|


__Zadanie 10.__ <br>
Do kodu z zadania 9 dodaj odczytanie nagłówków.<br>
Usuń puste wartości z nagłówków.<br>
Możesz użyć: filter(None, naglowki)

In [49]:
dane = []
with open(sciezka+r"/6. Pliki/MARKI_2019_05.csv", 'r') as plik:
    naglowki = plik.readline().rstrip().split('|')
    naglowki = list(filter(None, naglowki))
    print(naglowki)
    for l in plik:
        print(l, end = "")
        

['Marka', 'Liczba']
VOLKSWAGEN|15471|
OPEL|12410|
FORD|10590|
RENAULT|9683|
TOYOTA|9153|
AUDI|8248|
SKODA|7889|
BMW|7615|
MERCEDES-BENZ|6611|
PEUGEOT|6568|


__Zadanie 11.__ <br>
Utwórz słowniki i dodaj do listy jak: dane = [{'Marka':'VOLKSWAGEN, 'Liczba':15471}, ...]

In [11]:
dane = []
with open(sciezka+r"/6. Pliki/MARKI_2019_05.csv", 'r') as plik:
    naglowki = plik.readline().rstrip().split('|')
    naglowki = list(filter(None, naglowki))
    for l in plik:
        l = l.rstrip().split('|')
        l = list(filter(None, l))
        slownik = {}
        for k, v in zip(naglowki, l):
            slownik[k] = v
        print(slownik)
        dane.append(slownik)
print('-----------------------------------------')
print(dane)

{'Marka': 'VOLKSWAGEN', 'Liczba': '15471'}
{'Marka': 'OPEL', 'Liczba': '12410'}
{'Marka': 'FORD', 'Liczba': '10590'}
{'Marka': 'RENAULT', 'Liczba': '9683'}
{'Marka': 'TOYOTA', 'Liczba': '9153'}
{'Marka': 'AUDI', 'Liczba': '8248'}
{'Marka': 'SKODA', 'Liczba': '7889'}
{'Marka': 'BMW', 'Liczba': '7615'}
{'Marka': 'MERCEDES-BENZ', 'Liczba': '6611'}
{'Marka': 'PEUGEOT', 'Liczba': '6568'}
-----------------------------------------
[{'Marka': 'VOLKSWAGEN', 'Liczba': '15471'}, {'Marka': 'OPEL', 'Liczba': '12410'}, {'Marka': 'FORD', 'Liczba': '10590'}, {'Marka': 'RENAULT', 'Liczba': '9683'}, {'Marka': 'TOYOTA', 'Liczba': '9153'}, {'Marka': 'AUDI', 'Liczba': '8248'}, {'Marka': 'SKODA', 'Liczba': '7889'}, {'Marka': 'BMW', 'Liczba': '7615'}, {'Marka': 'MERCEDES-BENZ', 'Liczba': '6611'}, {'Marka': 'PEUGEOT', 'Liczba': '6568'}]


__Zadanie 12.__ <br>
Zapisz do pliku binarnego słownik z zadania 11

In [13]:
with open(sciezka+r"/6. Pliki/samochody.pickle", 'wb') as file:
    pickle.dump(dane, file)

__Zadanie 13.__ <br>
Odczytaj z pliku binarnego plik zapisany w zadaniu 11

In [15]:
with open(sciezka+r"/6. Pliki/samochody.pickle", 'rb') as file:
    samochody = pickle.load(file)
print(samochody)

[{'Marka': 'VOLKSWAGEN', 'Liczba': '15471'}, {'Marka': 'OPEL', 'Liczba': '12410'}, {'Marka': 'FORD', 'Liczba': '10590'}, {'Marka': 'RENAULT', 'Liczba': '9683'}, {'Marka': 'TOYOTA', 'Liczba': '9153'}, {'Marka': 'AUDI', 'Liczba': '8248'}, {'Marka': 'SKODA', 'Liczba': '7889'}, {'Marka': 'BMW', 'Liczba': '7615'}, {'Marka': 'MERCEDES-BENZ', 'Liczba': '6611'}, {'Marka': 'PEUGEOT', 'Liczba': '6568'}]
